src/main/java/core/basesyntax/MyLinkedList.java

```diff
@@ -3,48 +3,160 @@
 import java.util.List;
 
 public class MyLinkedList<T> implements MyLinkedListInterface<T> {
+    private Node<T> head; //початок
+    private Node<T> tail; //кінець
+    private int size = 0;
+
+    private static class Node<T> {
+        private T value;
+        private Node<T> next;
+        private Node<T> prev;
+
+        public Node(Node<T> prev, T value, Node<T> next) {
+            this.value = value;
+            this.next = next;
+            this.prev = prev;
+        }
+    }
+
     @Override
     public boolean add(T value) {
-        return false;
+        if (size == 0) {
+            head = new Node<>(null, value, null);
+            tail = head;
+            size++;
+            return true;
+        }
+        Node<T> lastTail = tail;
+        Node<T> newTail = new Node<>(lastTail, value, null);
+        lastTail.next = newTail;
+        tail = newTail;
+        size++;
+        return true;
     }
 
     @Override
     public void add(T value, int index) {
-
+        if (index > size && size != 0 || index < 0) {
+            throw new ArrayIndexOutOfBoundsException("Houston we have a BUG!");
+        }
+        if (index == size) {
+            add(value);
+            return;
+        }
+        Node<T> currentNext = search(index);
+        Node<T> currentPrev = currentNext.prev;
+        Node<T> newTail = new Node<>(currentPrev, value, currentNext);
+        if (currentPrev != null) {
+            currentPrev.next = newTail;
+        }
+        currentNext.prev = newTail;
+        if (index == 0) {
+            head = newTail;
+        }
+        size++;
     }
 
     @Override
     public boolean addAll(List<T> list) {
-        return false;
+        for (T t : list) {
+            add(t);
+        }
+        return true;
     }
 
     @Override
     public T get(int index) {
-        return null;
+        return search(index).value;
     }
 
     @Override
     public T set(T value, int index) {
-        return null;
+        checkIndex(index);
+        Node<T> current = search(index);
+        T currentOld = current.value;
+        current.value = value;
+        return currentOld;
     }
 
     @Override
     public T remove(int index) {
-        return null;
+        checkIndex(index);
+        Node<T> current = search(index);
+        Node<T> currentNext = current.next;
+        Node<T> currentPrev = current.prev;
+        if (index == 0) {
+            head = currentNext;
+        }
+        if (index == size - 1) {
+            tail = currentPrev;
+        }
+        if (currentPrev != null) {
+            currentPrev.next = currentNext;
+        }
+        if (currentNext != null) {
+            currentNext.prev = currentPrev;
+        }
```

Comment:
this code duplicates in remove(T object) let's move it to unlink(Node) method and call it in both removes

Reply:
done